File: examples/go-gotest/tables_and_docstrings.steps.go
package example

import (
	"strings"

	"github.com/varar-dev/varar-go/varar"
)

func registerTablesAndDocstrings(s *varar.Steps[Ctx]) {
	// A whole-table slot arrives as rows of cells and the computed table goes
	// back as rows keyed by column.
	s.Sensor("Uppercase each one:", func(ctx Ctx, table [][]string) ([]map[string]string, error) {
		out := []map[string]string{}
		for _, row := range table[1:] { // skip the header row
			out = append(out, map[string]string{"before": row[0], "after": strings.ToUpper(row[0])})
		}
		return out, nil
	})

	// Two slots — the {word} capture and the trailing doc string — so two
	// strings in, two strings out, compared positionally.
	s.Sensor("Greet {word}:", func(ctx Ctx, name, doc string) (string, string, error) {
		return name, "Hello, " + name + "!\n", nil
	})
}

File: examples/csharp-vstest/steps/tables-and-docstrings.steps.cs
using Varar;
using Varar.Core;
using static Varar.Example.StepHelpers;

namespace Varar.Example;

public static class TablesAndDocstringsSteps
{
    public static void Register(Steps s)
    {
        // A whole table is handed to the step; it returns the computed rows and Varar checks each cell.
        s.Sensor("Uppercase each one:", (state, table) =>
        {
            var rows = ((VList)table).Items;
            var computed = rows.Skip(1).Select(row =>
            {
                var before = AsStr(((VList)row).Items[0]);
                return VMap(("before", Value.Of(before)), ("after", Value.Of(before.ToUpperInvariant())));
            });
            return Value.List(computed);
        });

        // Two slots: the {word} and the trailing doc string; the return is [word, produced text].
        s.Sensor("Greet {word}:", (state, name, doc) =>
        {
            var n = name.AsString();
            return Value.List([Value.Of(n), Value.Of($"Hello, {n}!\n")]);
        });
    }
}
